internal/helper/diag.go

```diff
@@ -0,0 +1,86 @@
+package helper
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+
+	"github.com/honeycombio/terraform-provider-honeycombio/client"
+)
+
+// DetailedErrorDiagnostic is a Diagnostic which nicely wraps a client.DetailedError
+type DetailedErrorDiagnostic struct {
+	summary string
+	e       *client.DetailedError
+}
+
+// compile-time check that DetailedErrorDiagnostic implements diag.Diagnostic
+var _ diag.Diagnostic = DetailedErrorDiagnostic{}
+
+// NewDetailedErrorDiagnostic creates a new DetailedErrorDiagnostic
+// taking a context-specific summary of the action and a DetailedError.
+func NewDetailedErrorDiagnostic(summary string, e *client.DetailedError) DetailedErrorDiagnostic {
+	return DetailedErrorDiagnostic{
+		e:       e,
+		summary: summary,
+	}
+}
+
+// AddDiagnosticOnError is a helper function which will take an error
+// and a context-specific summary of the action.
+//
+// If err is nil, this function will return false.
+// If err is non-nil, this function will return ture and take one the following actions:
+//   - if it is a client.DetailedError, a DetailedErrorDiagnostic will be added to diag.
+//   - otherwise a generic error diagnostic will be added to diag.
+func AddDiagnosticOnError(diag *diag.Diagnostics, summary string, err error) bool {
```

Comment:
I attempted to come up with a signature for this helper that could expose the 404 case needed to be handled on all the `Read` methods for resources and didn't love anything I came up with.

Reply:
the best thing I can think of would be passing in a callback func to the function, like

```golang
trigger, err := r.client.Triggers.Get(ctx, state.Dataset.ValueString(), state.ID.ValueString())
if helper.AddDiagnosticOnError(&resp.Diagnostics, "Creating Honeycomb Trigger", err, func(detailedErr *client.DetailedError) bool {
	if detailedErr.IsNotFound() {
                // not super sure if this works with Go's closures like it would if this was JS
		resp.State.RemoveResource(ctx)
		return false
	}
	return true
}) {
  return
}

// ...

func AddDiagnosticOnError(diag *diag.Diagnostics, summary string, err error, callback func(detailedErr *client.DetailedError) bool) bool {
	if err == nil {
		return false
	}

	var detailedErr *client.DetailedError
	if errors.As(err, &detailedErr) {
		// the return val of callback determines if we try to append the error, letting the caller decide
		if callback(detailedErr) {			
			diag.Append(DetailedErrorDiagnostic{
				summary: "Error " + summary,
				e:       detailedErr,
			})
		}
	} else {
		diag.AddError("Error "+summary, err.Error())
	}
	return true
}
```

...but I don't think it makes it much clearer at all. what you have seems like a great improvement already!